File: bindings/python/examples/lgamma_bench.py
"""lgamma_bench.py — treeweave vs scipy.special.gammaln (log-Gamma).

The Python member of the cross-language lgamma benchmark family (see
examples/c++/lgamma_bench.cpp for the rationale). log-Gamma is fit on [3, 50) —
smooth, positive, monotone, so relative error is well defined — with treeweave's
default RelativeMax tolerance. The fit callback is the scalar ``math.lgamma``;
the throughput baseline is the vectorized ``scipy.special.gammaln`` (the fair
comparison against treeweave's vectorized batch path).

Requires NumPy and SciPy:  pip install numpy scipy
"""

import math
import time

import numpy as np
from scipy.special import gammaln  # vectorized log-Gamma baseline

import treeweave

a, b = 3.0, 50.0
# treeweave probes the scalar math.lgamma at Chebyshev nodes during the fit.
approx = treeweave.fit(lambda x: math.lgamma(x[0]), a, b, tol=1e-10)
print(approx)

n = 1_000_000
rng = np.random.default_rng(7)
xs = rng.uniform(a, b, n)

# --- accuracy vs the library -------------------------------------------------
yhat = approx(xs)
yref = gammaln(xs)
max_rel = float(np.max(np.abs(yhat - yref) / np.abs(yref)))

# --- throughput: treeweave vs scipy.special.gammaln --------------------------
approx(xs)  # warm-up
t0 = time.perf_counter()
r = approx(xs)
tw_s = time.perf_counter() - t0

gammaln(xs)  # warm-up
t0 = time.perf_counter()
ell = gammaln(xs)
lib_s = time.perf_counter() - t0

assert np.isfinite(r.sum() + ell.sum())  # keep the timed calls live

print(f"lgamma fit on [{a:.1f}, {b:.1f}), relative tol 1e-10")
print(f"  max rel err: {max_rel:.3e}")
print(f"  treeweave:  {n / (tw_s * 1e6):.1f} Mevals/s")
print(f"  library (scipy.special.gammaln): {n / (lib_s * 1e6):.1f} Mevals/s")
print(f"  speedup: {lib_s / tw_s:.2f}x")

File: bindings/matlab/examples/lgamma_bench.m
% lgamma_bench.m — treeweave vs the native gammaln (log-Gamma).
%
% The MATLAB/Octave member of the cross-language lgamma benchmark family (see
% examples/c++/lgamma_bench.cpp for the rationale). log-Gamma is fit on [3, 50)
% — smooth, positive, monotone, so relative error is well defined — with
% treeweave's default RelativeMax tolerance, then compared to the built-in
% gammaln (available in both MATLAB and Octave) on max relative error,
% throughput, and speedup.
%
% Before running: the build tree (treeweave_mex + the tw_*.m stubs) must be on
% the path. This script adds only the source dir (treeweave.m classdef).
addpath(fileparts(fileparts(mfilename('fullpath'))));   % bindings/matlab (treeweave.m)

a = 3;
b = 50;
obj = treeweave(@(x) gammaln(x(1)), [a], [b], 1e-10, 'dim', 1, 'out_dim', 1);

n = 1e6;
X = a + (b - a) * rand(n, 1);

% --- accuracy vs the library --------------------------------------------------
Yhat    = obj.eval(X);
Yref    = gammaln(X);
max_rel = max(abs(Yhat - Yref) ./ abs(Yref));

% --- throughput: treeweave vs gammaln ----------------------------------------
obj.eval(X);                    % warm-up
t = tic; Yhat = obj.eval(X); tw_s = toc(t);
t = tic; Yref = gammaln(X);  lib_s = toc(t);

fprintf('lgamma fit on [%.1f, %.1f), relative tol %.0e\n', a, b, 1e-10);
fprintf('  max rel err: %.3e\n', max_rel);
fprintf('  treeweave:  %.1f Mevals/s\n', n / (tw_s * 1e6));
fprintf('  library: %.1f Mevals/s\n', n / (lib_s * 1e6));
fprintf('  speedup: %.2fx\n', lib_s / tw_s);

delete(obj);
